feat: Make base read and write state classes more customizable

bilyana-gospodinova opened the pull request:
**Description**:
This PR modifies `ReadableKVStateBase` and `WritableKVStateBase` classes in order to make the read and write cache more customizable as a structure. This change is needed for the mirror node team in order to substitute our own scoped value based cache for concurrency reasons with historical state.

Update: 
_Current MN implementation_
We have our custom implementation of the `State` interface, which is a singleton bean and implements the required methods. For each model(Account, Token, TokenRels, etc) we have a separate class that extends the `ReadableKVStateBase`, reads the data from the DB and makes some transformations/adaptations of the model. Each of these classes is stateless, so it is a singleton as well. Each transaction has a context, which is a scoped value variable and in there we keep some data about the current transaction.

_Design specifics_
One of the requirements for the mirror node is that it can query historical data and that it can process multiple transactions concurrently at the same time.

_The issue_
With the current implementation there is a single read cache in the `ReadableKVStateBase` and a single modifications cache in `WritableKVStateBase`. So we can have a case where:
1. Transaction T1 reads account A against the latest block and puts it in the cache.
2. Transaction T2 reads account A against a historical block where account A does not exist yet and since the cache is shared, A is returned instead of null.
3. Other concurrent modifications of the cache occur as well. In any case you can't know if the data in the cache is for the current block that you search for.

_The possible solution_
We can enhance our transaction context and keep in there a tiny readable and writable cache that is relevant only for the current transaction. At the end of the transaction execution this context is disposed, so the cache will be cleared and also every transaction will be able to see only its own cache. For this purpose we need to override the current private maps in the `ReadableKVStateBase` and `WritableKVStateBase` classes and pass our own data structures from the transaction context. This was tested and works as expected in this PR - https://github.com/hashgraph/hedera-mirror-node/pull/10306. We need to push some of the required changes in here as well, so we can remove the copied classes.

Fixes https://github.com/hashgraph/hedera-services/issues/17818

steven-sheehy commented on platform-sdk/swirlds-state-api/src/main/java/com/swirlds/state/spi/ReadableKVStateBase.java:
```diff
@@ -138,6 +138,14 @@ protected final void markRead(@NonNull K key, @Nullable V value) {
+     * Returns the underlying map of all read values from the data source.
+     * @return a cache of all values read from this {@link ReadableKVState}.
+     */
+    protected ConcurrentMap<K, V> getReadCache() {
```
Doesn't this and the modifications need to be a regular map for our implementation using scoped value?

bilyana-gospodinova replied:
Yes, a regular map is enough for our case. I updated the return type.

steven-sheehy reviewed (approved):
LGTM

rbair23 commented:
Can you give some more detail on the requirement? I'm uncomfortable with exposing internal state.

rbair23 commented on platform-sdk/swirlds-state-api/src/main/java/com/swirlds/state/spi/ReadableKVStateBase.java:
```diff
@@ -138,6 +138,14 @@ protected final void markRead(@NonNull K key, @Nullable V value) {
+     * Returns the underlying map of all read values from the data source.
+     * @return a cache of all values read from this {@link ReadableKVState}.
+     */
+    protected Map<K, V> getReadCache() {
```
Should this at least be final?

bilyana-gospodinova replied:
It is not final as _we_ need to override it. I will update the description of this PR with more information about our case.

anthony-swirldslabs replied:
In that case, I suggest to annotate the method return value as `@NonNull` and add a note about this into the javadoc for this method, so as to ensure the overridden versions never return `null`. This is because the code above never checks the return value and just blindly chain-calls other methods from the cache instance. We don't want NPEs to pop up there.

rbair23 commented on platform-sdk/swirlds-state-api/src/main/java/com/swirlds/state/spi/WritableKVStateBase.java:
```diff
@@ -218,6 +219,14 @@ public long size() {
+     * Returns a map of all modified values buffered in this mutable state.
+     * @return map of all modified values
+     */
+    protected Map<K, V> getModifications() {
```
Should this at least be final? If not, someone can override and return something random.

bilyana-gospodinova replied:
It is not final as _we_ need to override it. I will update the description of this PR with more information about our case.

anthony-swirldslabs replied:
Similar to the above, please annotate the return value as `@NonNull` and add a note to the javadoc about this. For the exact same reason as above.

bilyana-gospodinova replied:
Thanks, comment addressed.

rbair23 commented:
Wouldn't it be better to pass the map to use into a secondary (protected) constructor rather than exposing and overwriting a method like this?

bilyana-gospodinova commented:
I updated the description of the PR with more information about our case and why we need this change.

anthony-swirldslabs reviewed (commented):
Looks good to me overall. But I added a couple comments. Please take a look.

anthony-swirldslabs reviewed (approved):
Thanks for addressing the comments. Functionally, the fix looks good to me.

However, I share the same sentiment as @rbair23 : perhaps, from the API perspective, it might be cleaner to pass in a cache instance to the constructor of the class rather than override methods in it.

I'll let you and @rbair23 (and other reviewers) finalize this decision.

bilyana-gospodinova commented:
I agree that it would be cleaner if we pass the needed collections to a constructor. However, this would not fully work for our case since our custom readable states are singletons and we would need to recreate them for each transaction just in order to pass a different map to a constructor. This way we would need to dispose a lot of objects and this will not be very optimal.

anthony-swirldslabs commented:
> I agree that it would be cleaner if we pass the needed collections to a constructor. However, this would not fully work for our case since our custom readable states are singletons and we would need to recreate them for each transaction just in order to pass a different map to a constructor. This way we would need to dispose a lot of objects and this will not be very optimal.

I'm unaware of the complexity of your use case or any specific details, so this suggestion may not be reasonable. With this in mind, I see that the cache here is an object that implements the `Map` interface. So in theory you can implement this `Map` interface in a way that supports the functionality that you're currently implementing in your overridden `getReadCache()` methods. This approach could allow you to pass "collections" (aka implementations of the Map interface - they can be collections with additional logic on top of them) to the constructor and not require overriding any methods. In other words, you'd move the getReadCache() logic into the cache implementation itself. Again, not sure if this approach is reasonable in your particular use-case, but it seems viable purely from the technical/Java perspective.

bilyana-gospodinova commented:
Thans, @anthony-swirldslabs, I addressed your comments and I opened a new PR from a fork - https://github.com/hiero-ledger/hiero-consensus-node/pull/17943. Closing this PR since I no longer have write access to this repository and I can't update the PR. Would you please update the new PR with the required assignments, milestones, labels, etc?
